keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -417,14 +483,24 @@ affect the perceived performance.
 3. As releases increment, the probablity of a WebSocket enabled `kubectl` communicating
 with an older non-WebSocket enabled API Server decreases.
 
-### Proposal: `K8s-Websocket-Protocol: stream-translate` Header
+### Proposal: New `RemoteCommand` Sub-Protocol Version - `v5.channel.k8s.io`
+
+The latest RemoteCommand version does not address an important protocol feature--a
+stream `CLOSE` signal. In order to communicate to another endpoint that the current
+stream of sending data is complete, a `CLOSE` signal is necessary. This problem
+currently arises when sending data over the STDIN stream, and it is more fully described
+in the following issue: [exec over web sockets protocol is flawed](https://github.com/kubernetes/kubernetes/issues/89899).
+A new RemoteCommand version (`v5.channel.k8s.io`) adds this `CLOSE` signal.
 
-In addition to the current SPDY-based clients, there are other current WebSocket clients,
-including a javascript/browser-based client. In order to distinguish these older
-WebSocket clients from the new stream-translated WebSocket clients, we propose adding
-a new header `K8s-Websocket-Protocol: stream-translate`. As described further in the
-next `Proposal` section, this header allows newer clients to delegate to the
-`StreamTranslatorProxy` to translate WebSockets data messages to SPDY.
+### Proposal: New `PortForward` Sub-Protocol Version - `v2.portforward.k8s.io`
```

Comment:
We'll have a limit on the number of ports one invocation can handle, right? I do not know if this limit existed before and it will be very high > 100 ports IIUIC, but I think is worth mentioning and add validation to the kubectl commands

Reply:
yes, we can change the limits by determining how many first bits will be allocated but in any case there will be some kind of a limit and I agree with you that mentioning this would be beneficial.

Reply:
Talked with @seans3 today and we need to maintain parity with the spdy implementation in terms of number of supported ports and streams to be able to default to using the websocket impl. This means the v5 will have to allow more than a single byte for the stream ID.

This could be done with a variable-length encoding to keep the typical case (low single digits of streams) efficient.

Reply:
Within the description of the next PortForward subprotocol (`v2.portforward.k8s.io`), I have added that we will create a variable length stream id generation mechanism allowing for up to 2^31 (2,147,483,648) streams identifiers. This will ensure the websocket version has parity with the current SPDY implementation when it comes to the number of streams supported.

**NOTE** We will need to change the `CLOSE` signal value from current 255 to 127, to allow for the first bit to signify more bytes in the variable length encoding for stream id.

Reply:
Sounds like a great idea to simplify and unify the code base because remotecommand's static stream id mechanism does not work for portforward and it brings about unnecessary code duplications. In addition to that one protocol covers both seems more maintainable.